File: C++/MergeIntervals.cc
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

struct Interval {
    int start;
    int end;
    Interval() : start(0), end(0) {}
    Interval(int s, int e) : start(s), end(e) {}
};

bool cmpIntervals(const Interval& interval1, const Interval& interval2)
{
    return interval1.start < interval2.start;
}

//Non-in-place version
vector<Interval> merge0(vector<Interval> &intervals) {
    vector<Interval> sortedIntervals = intervals;
    sort(sortedIntervals.begin(), sortedIntervals.end(), cmpIntervals);
    vector<Interval> result;

    for (int i = 0; i < sortedIntervals.size(); ++i)
    {
        if ((result.empty()) || (result.back().end < sortedIntervals[i].start))
            result.push_back(sortedIntervals[i]);
        else {
            result.back().end = max(result.back().end, sortedIntervals[i].end);
        }
    }

    return result;
}

//See here: https://oj.leetcode.com/discuss/7415/do-you-have-a-better-solution
//for in-place version
vector<Interval> merge(vector<Interval> &intervals) {
    sort(intervals.begin(), intervals.end(), cmpIntervals);
    int t = 0;
    for (int i = 1; i < intervals.size(); ++i)
    {
        if (intervals[i].start > intervals[t].end) {
            intervals[++t] = intervals[i];
        } else {
            intervals[t].end = max(intervals[t].end, intervals[i].end);
        }
    }
    
    if (!intervals.empty())
        intervals.erase(intervals.begin()+t+1, intervals.end());
        
    return intervals;
        
}

int main(int argc, char** argv)
{
    vector<Interval> intervals = {
        Interval(1, 3), 
        Interval(2, 6), 
        Interval(8, 10), 
        Interval(15, 18)
    };

    vector<Interval> result = merge(intervals);
    
    for (int i = 0; i < result.size(); ++i)
        cout << result[i].start << ", " << result[i].end << endl;

    return 0;
}

File: Java/MergeIntervals.java
/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
public class MergeIntervals {
    
    //Sort and then merge.
    //O(nlogn) time and O(n) space
    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> sortedIntervals = new ArrayList<>(intervals);
        Collections.sort(sortedIntervals, new Comparator<Interval>() {
                public int compare(Interval interval1, Interval interval2) {
                    return Integer.compare(interval1.start, interval2.start);
                }
            });
        List<Interval> result = new ArrayList<>();
        if (intervals.isEmpty()) {
            return result;
        }
        Interval prevInterval = new Interval(sortedIntervals.get(0).start, sortedIntervals.get(0).end);
        for (Interval interval : sortedIntervals) {
            if (interval.start > prevInterval.end) {
                result.add(prevInterval);
                prevInterval = interval;
            } else {
                prevInterval.end = Math.max(prevInterval.end, interval.end);
            }
        }
        result.add(prevInterval);
        return result;
    }

    //Better implementation
    public List<Interval> merge(List<Interval> intervals) {
        List<Interval> result = new ArrayList<>();
        if (intervals == null || intervals.isEmpty()) {
            return result;
        }
        Collections.sort(intervals, new Comparator<Interval>() {
                public int compare(Interval interval1, Interval interval2) {
                    return Integer.compare(interval1.start, interval2.start);
                }
        });
        for (Interval interval : intervals) {
            if (result.isEmpty() || result.get(result.size() - 1).end < interval.start) {
                result.add(interval);
            } else {
                Interval lastInterval = result.get(result.size() - 1);
                lastInterval.end = Math.max(lastInterval.end, interval.end);
            }
        }
        return result;
    }

}